File: benchmarks/mandelbrot/mandelbrot.py
import time

def main():
    width = 800
    height = 800
    max_iter = 100

    x_min = -2.0
    x_max = 1.0
    y_min = -1.5
    y_max = 1.5

    dx = (x_max - x_min) / width
    dy = (y_max - y_min) / height

    start = time.perf_counter()

    in_set = 0
    checksum = 0

    for py in range(height):
        cy = y_min + py * dy
        for px in range(width):
            cx = x_min + px * dx
            zx = 0.0
            zy = 0.0
            iter_count = 0
            while iter_count < max_iter:
                zx2 = zx * zx
                zy2 = zy * zy
                if zx2 + zy2 > 4.0:
                    break
                zy = 2.0 * zx * zy + cy
                zx = zx2 - zy2 + cx
                iter_count += 1
            if iter_count == max_iter:
                in_set += 1
            checksum += iter_count

    elapsed_ms = (time.perf_counter() - start) * 1000
    print(f"Time: {elapsed_ms:.3f} ms")
    print(f"In set: {in_set}")
    print(f"Checksum: {checksum}")

if __name__ == "__main__":
    main()

File: benchmarks/mandelbrot/mandelbrot.c
#include <stdio.h>
#include <time.h>

int main(void) {
    int width = 800;
    int height = 800;
    int max_iter = 100;

    double x_min = -2.0;
    double x_max = 1.0;
    double y_min = -1.5;
    double y_max = 1.5;

    double dx = (x_max - x_min) / (double)width;
    double dy = (y_max - y_min) / (double)height;

    struct timespec ts_start, ts_end;
    clock_gettime(CLOCK_MONOTONIC, &ts_start);

    int in_set = 0;
    long long checksum = 0;

    for (int py = 0; py < height; py++) {
        double cy = y_min + (double)py * dy;
        for (int px = 0; px < width; px++) {
            double cx = x_min + (double)px * dx;
            double zx = 0.0;
            double zy = 0.0;
            int iter = 0;
            while (iter < max_iter) {
                double zx2 = zx * zx;
                double zy2 = zy * zy;
                if (zx2 + zy2 > 4.0) {
                    break;
                }
                zy = 2.0 * zx * zy + cy;
                zx = zx2 - zy2 + cx;
                iter++;
            }
            if (iter == max_iter) {
                in_set++;
            }
            checksum += iter;
        }
    }

    clock_gettime(CLOCK_MONOTONIC, &ts_end);
    double elapsed_ms = (ts_end.tv_sec - ts_start.tv_sec) * 1000.0
                      + (ts_end.tv_nsec - ts_start.tv_nsec) / 1e6;
    printf("Time: %.3f ms\n", elapsed_ms);
    printf("In set: %d\n", in_set);
    printf("Checksum: %lld\n", checksum);

    return 0;
}
